### DISCUSSION TITLE
Warning - PHP Startup: Unable to load dynamic library

### DISCUSSION THREAD

**Discussion by vlodko** [shivammathur/homebrew-extensions#924] [Q&A]
I keep getting this warning in console:
```
$ php -v
PHP Warning: PHP Startup: Unable to load dynamic library '/usr/local/Cellar/php@7.1/7.1.33/lib/php/20160303/' - dlopen(/usr/local/Cellar/php@7.1/7.1.33/lib/php/20160303/, 9): no suitable image found. Did find:
/usr/local/Cellar/php@7.1/7.1.33/lib/php/20160303/: not a file
/usr/local/Cellar/php@7.1/7.1.33/lib/php/20160303: not a file in Unknown on line 0

PHP 7.1.33 (cli) (built: May 14 2021 06:31:00) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.1.33, Copyright (c) 1999-2018, by Zend Technologies
    with Xdebug v2.9.8, Copyright (c) 2002-2020, by Derick Rethans

```
Any thoughts?

**Comment by shivammathur** [comment]
@vlodko 
It seems you have a stray `extension=` line without the extension name or path in your `php.ini` configuration.
Run `php --ini` and check the files it mentions for `extension=`.

**Reply by pp-vlad** [reply]
thanks for the quick response @shivammathur !
I've found a fix in the discussions section of the main repo.
Just had to re-install php@7.1

**Comment by konekt-balaji** [comment]
I'm unable to run extensions on MacOS Monterey.. I get the below error

PHP Warning:  PHP Startup: Unable to load dynamic library '/opt/homebrew/opt/mongodb@7.4/mongodb.so' (tried: /opt/homebrew/opt/mongodb@7.4/mongodb.so (dlopen(/opt/homebrew/opt/mongodb@7.4/mongodb.so, 0x0009): tried: '/opt/homebrew/opt/mongodb@7.4/mongodb.so' (code signature in &lt;170534DE-52F1-36B7-84F5-3FDF2EDEF552&gt; '/opt/homebrew/Cellar/mongodb@7.4/1.11.1/mongodb.so' not valid for use in process: mapped file has no Team ID and is not a platform binary (signed with custom identity or adhoc?)), '/usr/lib/mongodb.so' (no such file), '/opt/homebrew/Cellar/mongodb@7.4/1.11.1/mongodb.so' (code signature in &lt;170534DE-52F1-36B7-84F5-3FDF2EDEF552&gt; '/opt/homebrew/Cellar/mongodb@7.4/1.11.1/mongodb.so' not valid for use in process: mapped file has no Team ID and is not a platform binary (signed with custom identity or adhoc?)), '/usr/lib/mongodb.so' (no such file)), /opt/homebrew/lib/php/pecl/20190902//opt/homebrew/opt/mongodb@7.4/mongodb.so.so (dlopen(/opt/homebrew/lib/php/pecl/20190902//opt/homebrew/opt/mongodb@7.4/mongo in Unknown on line 0

Can you help please?

**Reply by shivammathur** [reply]
Is it limited to mongodb, or fails for all extensions?

**Reply by konekt-balaji** [reply]
It fails for all extensions that i install from this repo and the error message is same for all

`not valid for use in process: mapped file has no Team ID and is not a platform binary (signed with custom identity or adhoc?)`

**Reply by shivammathur** [reply]
What is the source of PHP you are using?

**Reply by konekt-balaji** [reply]
I installed PHP from you repo too.. the command i used was 
`brew install shivammathur/php/php@7.4`

**Reply by konekt-balaji** [reply]
Btw.. i'm getting that error on the apache error log

**Reply by shivammathur** [reply]
ok, you might be using system Apache. please use Apache from brew as well.

https://getgrav.org/blog/macos-monterey-apache-multiple-php-versions

**Reply by konekt-balaji** [reply]
Yes indeed.. Let me try the apache from homebrew

**Reply by konekt-balaji** [reply]
@shivammathur I can confirm that using apache from homebrew solved the issue. Thank you for the prompt reply. Have a nice day :)

**Reply by Drex72** [reply]
Good day @shivammathur 
i'm having an issue with the imap extension
Warning: PHP Startup: Unable to load dynamic library 'imap'
it keeps throwing this and i'm frustrated at this point

**Reply by shivammathur** [reply]
@Drex72 Please create a new issue with details of your PHP version and how you installed the imap extension.

**Reply by Drex72** [reply]
Alright

**Comment by jovylle** [comment]
I do have an issue on installing a extension, how to add up the brew installed extension to my php in local

**Reply by shivammathur** [reply]
When you install the extension package using brew, it will be added to your PHP. Please run `php -m` to check the extensions installed and enabled. If you have more questions, please open a new discussion with all the details.

**Reply by jovylle** [reply]
Thank you shivammathur you had made this helpful repo.

it did worked now. On first it did show that it failed to install and it looks like it god conflict with other Mcrypt library instalation I made.
so I tried force brew link it, as it suggest, but did't install at all. can't find in php -m

then after few hours of hopeless searching. I accedentally found the Mcrypt location and remove all of then then re install the extension. And it did worked perfectly without conlfic and finaly show on on php -m

**Comment by Drex72** [comment]
@shivammathur 
So i'm using php 8.2 and i installed imap with homebrew
i installed the kim-api because i'm using version 8.2
but it still doesnt work

<img width="1539" alt="Screenshot 2023-10-10 at 9 30 58 AM" src="https://github.com/shivammathur/homebrew-extensions/assets/90635055/b850d1df-4132-4884-ae8b-d76ede56d324">

it just keeps on showing this

PHP Warning:  PHP Startup: Unable to load dynamic library 'imap' (tried: /usr/local/lib/php/pecl/20220829/imap (dlopen(/usr/local/lib/php/pecl/20220829/imap, 0x0009): tried: '/usr/local/lib/php/pecl/20220829/imap' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/local/lib/php/pecl/20220829/imap' (no such file), '/usr/local/lib/php/pecl/20220829/imap' (no such file)), /usr/local/lib/php/pecl/20220829/imap.so (dlopen(/usr/local/lib/php/pecl/20220829/imap.so, 0x0009): tried: '/usr/local/lib/php/pecl/20220829/imap.so' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/local/lib/php/pecl/20220829/imap.so' (no such file), '/usr/local/lib/php/pecl/20220829/imap.so' (no such file))) in Unknown on line 0

**Reply by Drex72** [reply]
I've tried everything
i've degraded back to php@7.1, i've upgraded
everything and it still doesnt work

**Reply by jovylle** [reply]
eyyy, try now PHP 5.6

**Reply by Drex72** [reply]
Alright
so that would be delete my current php 
and reinstall the php version 5.6

**Reply by jovylle** [reply]
Just keep trying, I fixed mine on above comment.

**Reply by Drex72** [reply]
Alright, thank you
sorry, quick question
do i install it by brew install php@5.6
or do i install it from shivammathur

**Reply by jovylle** [reply]
right I remembered, installing that versions is only possible though shivammathur. hhmm if you don't require 5.6 please try another method.

**Reply by Drex72** [reply]
Just installed version 5.6
So what next

**Reply by Drex72** [reply]
Ijust finished installing version 5.6 and i removed the semicolon from the extension=php_imap.dll in my php.ini file
it then throws this
PHP Warning:  PHP Startup: Unable to load dynamic library '/usr/local/lib/php/pecl/20131226/php_imap.dll' - dlopen(/usr/local/lib/php/pecl/20131226/php_imap.dll, 0x0009): tried: '/usr/local/lib/php/pecl/20131226/php_imap.dll' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/local/lib/php/pecl/20131226/php_imap.dll' (no such file), '/usr/local/lib/php/pecl/20131226/php_imap.dll' (no such file) in Unknown on line 0

Warning: PHP Startup: Unable to load dynamic library '/usr/local/lib/php/pecl/20131226/php_imap.dll' - dlopen(/usr/local/lib/php/pecl/20131226/php_imap.dll, 0x0009): tried: '/usr/local/lib/php/pecl/20131226/php_imap.dll' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/local/lib/php/pecl/20131226/php_imap.dll' (no such file), '/usr/local/lib/php/pecl/20131226/php_imap.dll' (no such file) in Unknown on line 0

**Reply by shivammathur** [reply]
The extension for extension files on macos is `.so` and not `.dll`. Please update the entry in your `php.ini`.

**Reply by Drex72** [reply]
alright
let me do that

**Reply by Drex72** [reply]
same error
PHP Warning:  PHP Startup: Unable to load dynamic library '/usr/local/lib/php/pecl/20131226/php_imap.so' - dlopen(/usr/local/lib/php/pecl/20131226/php_imap.so, 0x0009): tried: '/usr/local/lib/php/pecl/20131226/php_imap.so' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/local/lib/php/pecl/20131226/php_imap.so' (no such file), '/usr/local/lib/php/pecl/20131226/php_imap.so' (no such file) in Unknown on line 0

Warning: PHP Startup: Unable to load dynamic library '/usr/local/lib/php/pecl/20131226/php_imap.so' - dlopen(/usr/local/lib/php/pecl/20131226/php_imap.so, 0x0009): tried: '/usr/local/lib/php/pecl/20131226/php_imap.so' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/local/lib/php/pecl/20131226/php_imap.so' (no such file), '/usr/local/lib/php/pecl/20131226/php_imap.so' (no such file) in Unknown on line 0

**Reply by jovylle** [reply]
have you reopened terminal? or atleast express your effort(by telling more things not just "same error") since we know you're doing the effort of fixing this.

**Reply by Drex72** [reply]
Yh
i reopened the terminal and i still got the same error
so i have a legacy codebase where they used imap for email stuff and when i execute the code, it throws an error saying that the function is undefined, so i then found out that the imap extension was not enabled

**Reply by Drex72** [reply]
So i enabled it in my php.ini, that was when i started to get these errors
and i've tried all i've seen
i've reinstalled php multiple times 
i tried to use xampp to install php, didnt work out
nothing has been working so far basically
